### PR TITLE
Feature: add `geth` dump parsing

### PR DESCRIPTION
**Pull request by Tirka**

### CODE REVIEW

File: evm-utils/evm-state/src/storage/mod.rs [hunk 1]
```diff
@@ -627,6 +689,65 @@ impl Storage<OptimisticTransactionDBInner> {
...
+            },
+        ) in accounts
+        {
+            let account_state = AccountState {
+                nonce,
+                balance,
+                code: code.into(),
+            };
+
+            set_account_state(&mut state_updates, address, account_state);
```
Comment (gBaGu) [inline evm-utils/evm-state/src/storage/mod.rs]: Consider something like this to remove fn set_account_state
```suggestion
            state_updates
                .entry(address)
                .or_insert((Maybe::Nothing, HashMap::new()))
                .0 = Maybe::Just(account_state);
```

File: evm-utils/evm-state/src/storage/mod.rs [hunk 2]
```diff
@@ -627,6 +689,65 @@ impl Storage<OptimisticTransactionDBInner> {
...
+            ext_storage(&mut state_updates, address, storage);
```
Comment (gBaGu) [inline evm-utils/evm-state/src/storage/mod.rs]: The same is above, i think fn ext_storage is redundant considering it is created inside a function and only used once
```suggestion
            state_updates
                .entry(address)
                .or_insert_with(|| (Maybe::Just(AccountState::default()), HashMap::new()))
                .1.extend(storage);
```

**Review by gBaGu — COMMENTED** [review]
Is there a reason why serde was not used to read geth dump?

File: sdk/src/genesis_config.rs [hunk 3]
```diff
@@ -429,417 +523,575 @@ pub mod evm_genesis {
...
+        /// Streaming deserializer for key,value pair in json.
+        /// input format is following:
+        ///
+        /// line0:   { "state": {
+        /// line1-N: "0x....": {...},
+        /// lineN:   }}
+        ///
+        /// serde_json StreamDeserializer can only work with valid json Value. '"key":{object}' - is not a valid json Value.
+        ///
+        pub struct OpenEthereumAccountExtractor<R: std::io::Read> {
```
Comment (vldm) [inline sdk/src/genesis_config.rs:600]: No reason to keep bounds in structs

File: sdk/src/genesis_config.rs [hunk 4]
```diff
@@ -429,417 +523,575 @@ pub mod evm_genesis {
...
-                    format!("Trying to read header of evm state json file, and it is invalid, should be '{{ \"state\": {{' got: {}", buffer),
-                ));
+                Ok(Self { reader })
             }
+        }
 
-            Ok(Self {
-                reader: IoRead::new(reader),
-            })
+        impl<'a> OpenEthereumAccountExtractor<&'a [u8]> {
```
Comment (vldm) [inline sdk/src/genesis_config.rs:611]: why you need two impls?
Why not generic over reader?

File: sdk/src/genesis_config.rs (hunk 3 above)
Comment (Tirka) [reply sdk/src/genesis_config.rs:600]: `BufReader` requires `R` to be `std::io::Read`

File: sdk/src/genesis_config.rs [hunk 5]
```diff
@@ -220,29 +222,64 @@ impl GenesisConfig {
...
+        std::fs::create_dir_all(ledger_path)?;
+        if let Some(mut dump_extractor) = dump_extractor {
+            let mut storage = Storage::open_persistent(ledger_path, true).unwrap();
+            let mut state_root = evm_state::empty_trie_hash();
+
+            let mut chunk = vec![];
+            let mut chunk_heap_bytes: usize = 0;
+            let mut accounts_pointer: usize = 1;
+            let mut no_more_accounts = false;
+            loop {
```
Comment (vldm) [inline sdk/src/genesis_config.rs]: This method is really complex.
1. Big cycle
2. cycle is loop, with control flow in middle, so it hard to trace it.
3. A lot of mutable variables.

Can we split this loop into two cycles:

1. `chunker` that limit chunk by value size in bytes.
2. processor - that handle each chunk.

I also propose to move chunker to seperate function and reuse peekable iterator to make code more clean.
something like this should work: 
```rust
fn get_chunk(&mut Iterator<Item=Result<Account,_>>) -> Vec<Chunk>;

while dump_extractor.peek().is_some() {
   let chunk = get_chunk();
   log::info!(
                   "Adding {} accounts to storage",
                    chunk.len(),
                    );
                    state_root = storage.set_initial(mem::take(&mut chunk), state_root);
}

assert_eq!(state_root, self.evm_root_hash);
```

If you need a counter for how many accounts processed at time of calling set_initial (aka accounts_pointer), just you can use `accounts_processed += chunks.len()`

File: sdk/src/genesis_config.rs (hunk 4 above)
Comment (Tirka) [reply sdk/src/genesis_config.rs:611]: I reduced code duplication by extracting common parts to generic implementation, but I'd like to keep these impls as thin wrappers for better expressiveness of public API. Commit hash: [ef0d522](https://github.com/velas/velas-chain/pull/417/commits/ef0d5228d70123d8099846df8b52678ce0b422ef)

File: evm-utils/evm-state/src/storage/mod.rs (hunk 2 above)
Comment (Tirka) [reply evm-utils/evm-state/src/storage/mod.rs]: Good call! Fixed at [b30ef6b](https://github.com/velas/velas-chain/pull/417/commits/b30ef6b4df950e1f2aab75f104b86e784f635a42)

File: sdk/src/genesis_config.rs (hunk 5 above)
Comment (Tirka) [reply sdk/src/genesis_config.rs]: Function splitted in two sequential subtasks: 
1. batching (with help of `batching` from `itertools`)
2. writing into storage

Fixed at: [ec0b7ec](https://github.com/velas/velas-chain/pull/417/commits/ec0b7ec2d6c3e6e77671e71ece279a3a72b0614e)

**Review by vldm — APPROVED** [review]

**Pull request merged**